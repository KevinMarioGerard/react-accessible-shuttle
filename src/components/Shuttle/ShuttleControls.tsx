import * as React from 'react';
import { ShuttleContext } from './ShuttleContext';
import { SHUTTLE_CONTROL_TYPES } from './reducers/index';
import { XOR } from '../../utils/macros';

type ShuttleControlsProps = XOR<
    { children?: React.ReactNode | ((setShuttleState: Function) => React.ReactNode) },
    { render?: (setShuttleState: Function) => React.ReactNode }
>;

/**
 * ShuttleControls. Provide your own render function
 * to customize the buttons. The Shuttle `dispatch` method
 * is provided or you can use your own state reducer
 * as needed.
 *
 * @example
 * // Custom render controls
 * <Shuttle.Controls render={setShuttleState => (
 *      <CustomButton onClick={() => setShuttleState({ type: 'MOVE_ALL' })} />
 * )} />
 */
export const ShuttleControls = React.memo(({ children, render, ...rest }: ShuttleControlsProps) => {
    const { setShuttleState } = React.useContext(ShuttleContext);

    if (typeof children === 'function') {
        return children(setShuttleState);
    }

    if (typeof render === 'function') {
        return render(setShuttleState);
    }

    const moveAllFromSource = React.useCallback(() => {
        setShuttleState({
            type: SHUTTLE_CONTROL_TYPES.MOVE_ALL,
            from: 'source',
            to: 'target',
        });
    }, []);

    const moveAllFromTarget = React.useCallback(() => {
        setShuttleState({
            type: SHUTTLE_CONTROL_TYPES.MOVE_ALL,
            from: 'target',
            to: 'source',
        });
    }, []);

    const moveSelectedFromSource = React.useCallback(() => {
        setShuttleState({
            type: SHUTTLE_CONTROL_TYPES.MOVE_SELECTIONS,
            from: 'source',
            to: 'target',
        });
    }, []);

    const moveSelectedFromTarget = React.useCallback(() => {
        setShuttleState({
            type: SHUTTLE_CONTROL_TYPES.MOVE_SELECTIONS,
            from: 'target',
            to: 'source',
        });
    }, []);

    return (
        <div className="shuttle__controls" {...rest}>
            <button onClick={moveAllFromSource}>{'\u00BB'}</button>
            <button onClick={moveSelectedFromSource}>{'\u203A'}</button>
            <button onClick={moveSelectedFromTarget}>{'\u2039'}</button>
            <button onClick={moveAllFromTarget}>{'\u00AB'}</button>
        </div>
    );
});

ShuttleControls.displayName = 'Shuttle.Controls';
