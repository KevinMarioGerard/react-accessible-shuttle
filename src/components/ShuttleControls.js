// @flow

import * as React from 'react';
import { ShuttleContext } from './ShuttleContext';

type Props =
    | {| children: React.Node | (() => React.Node) |}
    | {| render: () => React.Node |};

/**
 * ShuttleControls. Provide your own render function
 * to customize the buttons. The Shuttle `dispatch` method
 * is provided or you can use your own state reducer
 * as needed.
 *
 * Exposed reducers:
 * 1. `MOVE_ALL_TO_TARGET`
 * 2. `MOVE_SELECTED_TO_TARGET`
 * 3. `MOVE_SELECTED_TO_SOURCE`
 * 4. `MOVE_ALL_TO_SOURCE`
 *
 * @example
 * // Custom render controls
 * <Shuttle.Controls render={dispatch => (
 *      <CustomButton onClick={() => dispatch({ type: 'MOVE_ALL_TO_TARGET' })} />
 * )} />
 */
function ShuttleControls({ children, render }: Props) {
    const { dispatch } = React.useContext(ShuttleContext);

    if (typeof children === 'function') {
        return children(dispatch);
    }

    if (typeof render === 'function') {
        return render(dispatch);
    }

    return (
        <div className="shuttle__controls">
            <button onClick={() => dispatch({ type: 'MOVE_ALL_TO_TARGET' })}>
                {'\u00BB'}
            </button>
            <button
                onClick={() => dispatch({ type: 'MOVE_SELECTED_TO_TARGET' })}>
                {'\u203A'}
            </button>
            <button
                onClick={() => dispatch({ type: 'MOVE_SELECTED_TO_SOURCE' })}>
                {'\u2039'}
            </button>
            <button onClick={() => dispatch({ type: 'MOVE_ALL_TO_SOURCE' })}>
                {'\u00AB'}
            </button>
        </div>
    );
}

ShuttleControls.displayName = 'Shuttle.Controls';

export default ShuttleControls;
