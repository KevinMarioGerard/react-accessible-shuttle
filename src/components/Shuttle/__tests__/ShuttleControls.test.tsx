import * as React from 'react';
import { render, cleanup } from '@testing-library/react';

import { ShuttleControls } from '../ShuttleControls';

afterEach(cleanup);

describe('ShuttleControls tests', () => {
    it('should allow customized rendering via child render function', () => {
        const { container } = render(
            <ShuttleControls>
                {({ moveAllFromSource, moveAllFromTarget }) => (
                    <>
                        <button onClick={moveAllFromSource}>Move All to Target</button>
                        <button onClick={moveAllFromTarget}>Move All to Source</button>
                    </>
                )}
            </ShuttleControls>
        );

        expect(container).toMatchSnapshot();
    });
});
