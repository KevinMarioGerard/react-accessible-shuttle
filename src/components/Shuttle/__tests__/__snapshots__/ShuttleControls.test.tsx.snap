// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShuttleControls tests should allow customized rendering via child render function 1`] = `
<div>
  <div
    class="shuttle__controls"
  >
    <button>
      Move All to Target
    </button>
    <button>
      Move All to Source
    </button>
  </div>
</div>
`;
